### PR TITLE
RavenDB-7059 implementation

### PR DESCRIPTION
**Pull request by myarichuk**

### CODE REVIEW

File: src/Raven.Server/Documents/Handlers/StatsHandler.cs [hunk 1]
```diff
@@ -18,14 +19,21 @@ public class StatsHandler : DatabaseRequestHandler
...
             {
                 var indexes = Database.IndexStore.GetIndexes().ToList();
                 var transformersCount = Database.TransformerStore.GetTransformersCount();
 
                 var stats = new DatabaseStatistics();
+                using (Database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext transactionOperationContext))
+                using (transactionOperationContext.OpenReadTransaction())
+                {
+                    var record = Database.ServerStore.Cluster.ReadDatabase(transactionOperationContext, Database.Name);
+                    stats.Identities = record.Identities;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Handlers/StatsHandler.cs]: Maybe this should go to a separate endpoint? Don't need to overload the DB stats.

File: src/Raven.Server/ServerWide/ClusterStateMachine.cs [hunk 2]
```diff
@@ -92,16 +92,36 @@ protected override void Apply(TransactionOperationContext context, BlittableJson
...
+                        nameof(UpdateClusterIdentityCommand),
+                        nameof(UpdateClusterIdentityCommand.Identities),
+                        out errorMessage))
+                    {
+                        NotifyLeaderAboutError(index, leader,
+                            new InvalidDataException(errorMessage));
+                        return;
+                    }
+                   
+                    goto case nameof(DeleteOngoingTaskCommand);
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ClusterStateMachine.cs]: Why are you deleting the task here? Why not use `break`?

File: src/Raven.Server/ServerWide/ServerStore.cs [hunk 3]
```diff
@@ -920,9 +920,9 @@ public DynamicJsonArray GetClusterErrors()
             return _engine.GetClusterErrorsFromLeader();
         }
 
-        public async Task<string> GenerateClusterIdentityAsync(string id, string databaseName)
+        public async Task<(long clusterEtag, string clusterId)> GenerateClusterIdentityAsync(string id, string databaseName)
```
Comment (ayende) [inline src/Raven.Server/ServerWide/ServerStore.cs]: Use PublicCasing

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 4]
```diff
@@ -232,14 +233,34 @@ public void Dispose()
...
+                public UpdateClusterIdentitiesCommand(Dictionary<string, long> identities, DocumentDatabase database)
+                {
+                    _identities = identities;
+                    _database = database;
+                }
+
+                public override int Execute(DocumentsOperationContext context)
+                {
+                    var (etag,_) = _database.ServerStore.SendToLeaderAsync(new UpdateClusterIdentityCommand(_database.Name,_identities)).Result;
+                    _database.ServerStore.WaitForCommitIndexChange(RachisConsensus.CommitIndexModification.GreaterOrEqual, etag).Wait();
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Why?

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 5]
```diff
@@ -232,14 +233,34 @@ public void Dispose()
...
+
+                public UpdateClusterIdentitiesCommand(Dictionary<string, long> identities, DocumentDatabase database)
+                {
+                    _identities = identities;
+                    _database = database;
+                }
+
+                public override int Execute(DocumentsOperationContext context)
+                {
+                    var (etag,_) = _database.ServerStore.SendToLeaderAsync(new UpdateClusterIdentityCommand(_database.Name,_identities)).Result;
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: You are _inside the transaction merger_, making a _raft consensus call_. That is not a good idea!

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs [hunk 6]
```diff
@@ -11,6 +11,6 @@ public enum DatabaseItemType
         RevisionDocuments = 1 << 1,
         Indexes = 1 << 2,
         Transformers = 1 << 3,
-        Identities = 1 << 4,
+        ClusterIdentities = 1 << 4
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: @myarichuk please open an issue that we need to support this block in 3.x (by skipping it). IIRC the 3.x smuggler throws if not supported property is encountered.

Comment (ayende) [reply src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: Why change the name of this anyway? There is no distinction

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs [hunk 7]
```diff
@@ -174,32 +174,45 @@ private void SkipType(DatabaseItemType type, SmugglerResult result)
...
-            using (var actions = _destination.Identities())
+            using (var clusterIdentityActions = _destination.ClusterIdentities())
             {
-                foreach (var kvp in _source.GetIdentities())
-                {
-                    _token.ThrowIfCancellationRequested();
-                    result.Identities.ReadCount++;
+                var sourceIdentities = _source.GetClusterIdentities();
+                WriteIdentities(result, sourceIdentities, clusterIdentityActions,
+                    () => result.ClusterIdentities.ReadCount++,
```
Comment (ppekrol) [inline src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: If we have one type of identities, then maybe we can get rid of the lambdas here (maybe revert back to the old code?)

File: src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs (hunk 6 above)
Comment (myarichuk) [reply src/Raven.Client/Documents/Smuggler/DatabaseItemType.cs]: hmm, no reason to change the name, I will change it back to Identities at the next PR

File: src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs (hunk 7 above)
Comment (myarichuk) [reply src/Raven.Server/Smuggler/Documents/DatabaseSmuggler.cs]: I will get rid of lambdas

File: src/Raven.Client/Documents/Operations/GetClusterIdentitiesOperation.cs [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Text;
+using Raven.Client.Documents.Conventions;
+using Raven.Client.Http;
+using Sparrow.Json;
+
+namespace Raven.Client.Documents.Operations
+{
+    public class GetClusterIdentitiesCommand : RavenCommand<Dictionary<string, long>>
```
Comment (ayende) [inline src/Raven.Client/Documents/Operations/GetClusterIdentitiesOperation.cs]: Rename to get database identities.

File: src/Raven.Server/Documents/Handlers/Debugging/IdentityDebugHandler.cs [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+    public class IdentityDebugHandler : DatabaseRequestHandler
+    {
+        [RavenAction("/databases/*/debug/identities", "GET")]
+        public Task GetIdentities()
+        {
+            using (Database.ServerStore.ContextPool.AllocateOperationContext(out TransactionOperationContext context))
+            using (context.OpenReadTransaction())
+            {
+                var record = Database.ServerStore.Cluster.ReadDatabase(context, Database.Name);
+                var identitiesAsJson = record.Identities.ToJson();
```
Comment (ayende) [inline src/Raven.Server/Documents/Handlers/Debugging/IdentityDebugHandler.cs]: What happens if we have a lot of identities, should we place a limit on that?

File: src/Raven.Server/ServerWide/Commands/UpdateClusterIdentityCommand.cs [hunk 10]
```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Client.Extensions;
+using Raven.Client.Server;
+using Sparrow.Json.Parsing;
+
+namespace Raven.Server.ServerWide.Commands
+{
+    public class UpdateClusterIdentityCommand : UpdateDatabaseCommand
```
Comment (ayende) [inline src/Raven.Server/ServerWide/Commands/UpdateClusterIdentityCommand.cs]: Rename class

File: src/Raven.Server/Smuggler/Documents/Data/ISmugglerSource.cs [hunk 11]
```diff
@@ -15,7 +15,7 @@ public interface ISmugglerSource
         IEnumerable<DocumentItem> GetRevisionDocuments(List<string> collectionsToExport, INewDocumentActions actions);
         IEnumerable<IndexDefinitionAndType> GetIndexes();
         IEnumerable<TransformerDefinition> GetTransformers();
-        IEnumerable<KeyValuePair<string, long>> GetIdentities();
+        IEnumerable<KeyValuePair<string, long>> GetClusterIdentities();
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/Data/ISmugglerSource.cs]: Rename method

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 12]
```diff
@@ -50,8 +52,8 @@ public IDocumentActions RevisionDocuments()
         {
             return new DatabaseDocumentActions(_database, _buildType, isRevision: true, log: _log);
         }
-
-        public IIdentityActions Identities()
+     
+        public IIdentityActions ClusterIdentities()
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: Rename method

File: src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs [hunk 13]
```diff
@@ -227,36 +226,11 @@ public void WriteIdentity(string key, long value)
...
-                public override int Execute(DocumentsOperationContext context)
-                {
-                    _database.DocumentsStorage.Identities.Update(context, _identities);
+                if (_identities.Count == 0)
+                    return;
 
-                    return 1;
-                }
+                //fire and forget, do not hold-up smuggler operations waiting for Raft command
+                _database.ServerStore.SendToLeaderAsync(new UpdateClusterIdentityCommand(_database.Name, _identities)).Wait();
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/DatabaseDestination.cs]: You _aren't_ doing fire and forget here.
And the smuggler is mostly async anyway. Move this behavior to FlushAsync()

File: src/Raven.Server/Smuggler/Documents/StreamDestination.cs [hunk 14]
```diff
@@ -60,11 +60,12 @@ public IDocumentActions RevisionDocuments()
             return new StreamDocumentActions(_writer, _context, _source, isRevision: true);
         }
 
-        public IIdentityActions Identities()
+        public IIdentityActions ClusterIdentities()
```
Comment (ayende) [inline src/Raven.Server/Smuggler/Documents/StreamDestination.cs]: Rename

**Review by ayende — CHANGES_REQUESTED** [review]

**Comment by ayende** [comment]
Manually merged

**Pull request closed without merging**